# Breeding-Value-Prediction

*Alternative Methods to Breeding Value Prediction in Loblolly Pine*

## Abstract

* Phenotypic variation in forest trees may be partitioned into genomic and environemntal compenets which are then used to estimate the heritability of traits of as the proportion of total phenotypic variation attributed to genetic variation.

* Applied tree breeding programs can use matrices of relationships, based either on recorded pedigrees in structured breeding populations or on genotypes of molecular genetic markers, to model genetic covariation among related individuals and predict genetic values for individuals for whom no phenotypic measurements are available. 

* ***This study tests the hypothesis that genetic covariation among individuals of similar genetic value will be reflected in shared patterns of gene expression.*** We collected gene expression data by high-throughput sequencing of RNA isolated from pooled seedlings from parents of known genetic value, and compared alternative approaches to data analysis to test this hypothesis.

## Background of samples

* LGEP

* EW

## Location of data


Data | Path | Notes
--- | --- | ---
**raw read files** 
*raw tar* | `/media/disk6/ARF/RNASEQ/rawreads/86kSalmon/` | `./EWtarfiles or ./LGEPtarfiles`
*raw fasta* | `/media/disk6/ARF/RNASEQ/rawreads/86kSalmon/` | `./EWfasta or ./LGEPfasta`
**trimmed and filtered read files** 
*EW* | `/media/disk6/ARF/RNASEQ/trimmedfiltreads/86k/EW/` | `./lane01 ... ./lane12`
*LGEP* | `/media/disk6/ARF/RNASEQ/trimmedfiltreads/86k/LGEP/` | `./lane01 ... ./lane18`
