# 1/29/18 SNP calling notes

*Raw tar files returned from GSL were extracted and names of folders were changed to correspond to a single pool (lane) run for each batch.  The .fastq.gz files for each library were QC'ed with fastqc and then adapters/contamination was removed using bbduk.  The filtered and trimmed reads are located at:

```linux
/media/disk6/ARF/RNASEQ/trimmedfiltreads
```

*The above path is the starting directory for processing SNPs.  Notes of lane extraction, fastqc, and filter/trim processing are on git at:

```linux
Breeding-Value-Prediction/disk6directory/RNASEQ/rawreads/processingnotes
```

*Below is a list of all steps. First each step will be practiced and once a final method for each step is determined a pipe command will be built for as many steps as possible.

## 1. Align .fastq.gz techincal reps to 86K transcriptome with bowtie

*Build bowtie2 index:

```linux
cd /media/disk6/ARF/RNASEQ
./software/bowtie/bowtie2-build -f ./references/Pita.86ktxptome.fasta Pita.86k
```

*Practice run bowtie2:

```linux
./software/bowtie/bowtie2 -p 60 -x ./references/Pta.86k -U ./trimmedfiltreads/EW/lane01/96_S12_L001_R1_001.fastq.gz -S out.sam
```
*The above outputs a .sam file that is uncompressed

*The below puts out a .bam file that is compressed. Note that when using samtools view "q" can be used to filter low quality reads. Should this be used?

```linux
while read file; do ./software/bowtie/bowtie2 -p 60 -x ./references/Pta.86k -U ./trimmedfiltreads/EW/lane01/$file | ./software/samtools-1.6/samtools view -@ 60 -b1 - -o ./snps/86k/EW/lane01/$file; done < ./rawreads/EWfasta/lane01/lane01.list
```

## 2. Convert sam file to bam
## 3. Merge technical rep bam files
## 4. Sort biological rep bam file
## 5. Mark/remove duplicates within bam files
## 6. Add RG & SM to each biological rep
## 7. Provide list of bams to freebayes and call SNPs
## 8. Filter VCF for no missing SNPs
