---
title: "Prediction of breeding values using LMM v2"
author: "Adam Festa"
date: "5/9/2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Script objective


```{r cars}
summary(cars)
```

## Load data

```{r pressure, echo=FALSE}
# Load counts
load("/mnt/media/disk6/ARF/RNASEQ/counts/86kSalmon/Step3_LMM_animal.RData")

# Load phenos
load("/mnt/media/disk6/ARF/RNASEQ/Breeding-Value-Prediction/disk6directory/resources/expt.dat.720.RData")
# Change factor to character and subset 3 columns
expt.dat.720$fam_id <- as.character(expt.dat.720$fam_id)
select.columns <- c("fam_id","animal_id","Volume","Height")

# Remove missing phenotypes and subset for phenos which have counts
fam.phenos <- unique(expt.dat.720[-which(is.na(expt.dat.720$Volume) == T),select.columns])
#fam.phenos <- fam.phenos[which(fam.phenos$fam_id %in% rownames(asreml.counts)),]
#rownames(fam.phenos) <- fam.phenos$fam_id


# Edit rownames of asreml.counts
 sub.names <- gsub(pattern = "animal_id, var = T)_",replacement = "",x = rownames(asreml.counts)[-c(1:94)])
 sub.names <- gsub("ped\\(", "", sub.names)
    
#rownames(asreml.counts) <- edit.names

# Subset each count data to contain only rows which have phenos
select.rows <- which(sub.names %in% as.character(fam.phenos$animal_id))

asreml.counts <- asreml.counts[select.rows,]

select.rows <- which(as.character(fam.phenos$animal_id) %in% sub.names)
fam.phenos <- fam.phenos[select.rows,]
```


## Summary stats

```{r}
sum.of.counts.per.gene <- apply(asreml.counts,2,sum)
sum.of.counts.per.fam <- apply(asreml.counts,1,sum)
hist(sum.of.counts.per.fam)
```

## Predictions

### Omic Kriging

```{r}
suppressPackageStartupMessages(library(OmicKriging))
```

#### All transcripts

```{r}
cor.mat <- cor(t(scale(log2(asreml.counts + 1),center=F)))
rownames(cor.mat) <- 1:190; colnames(cor.mat) <- 1:190
rownames(fam.phenos) <- 1:190
the.predictions <- sapply(1:57,function(each.fam){
  these <- which(as.character(fam.phenos$fam_id) %in% unique(as.character(fam.phenos$fam_id))[each.fam])
  test.fam <- these
  train.fam <- c(1:190)[-these]
  mean(okriging(idtest = test.fam,idtrain = train.fam,corlist = list(cor.mat),H2vec = c(.99),pheno = fam.phenos,phenoname = "Volume")[,2] - mean(fam.phenos$Volume[-these]))
})

cor(the.predictions,pheno.57$Volume)^2
plot(the.predictions,pheno.57$Volume)
```

#### BCV transcripts

```{r}
mean.cts <- apply(asreml.counts,2,mean)
sd.cts <- apply(asreml.counts,2,sd)
bcv.cts <- sd.cts/mean.cts
cor.mat <- cor(t(log2(asreml.counts[,which(bcv.cts > 1)] + 1)))
rownames(cor.mat) <- 1:190; colnames(cor.mat) <- 1:190

the.predictions <- sapply(1:57,function(each.fam){
  these <- which(as.character(fam.phenos$fam_id) %in% unique(as.character(fam.phenos$fam_id))[each.fam])
  test.fam <- these
  train.fam <- c(1:190)[-these]
  mean(okriging(idtest = test.fam,idtrain = train.fam,corlist = list(cor.mat),H2vec = c(.99),pheno = fam.phenos,phenoname = "Volume")[,2] - mean(fam.phenos$Volume[-these]))
})

cor(the.predictions,pheno.57$Volume)^2
plot(the.predictions,pheno.57$Volume)
```

#### Linear model regression

```{r}

p.vals <- mclapply(1:57,function(each.fam){
  test.fam <- each.fam
  train.fam <- c(1:57)[-each.fam]
  train.phenos <- fam.phenos$Volume[train.fam]
  sapply(1:ncol(asreml.counts),function(x){
    if(var(asreml.counts[train.fam,x]) > 0.001){
      summary(lm(fam.phenos$Volume[train.fam] ~ asreml.counts[train.fam,x]))[[4]][[8]] } else { NA }
    })
},mc.cores=30)
```

```{r}
the.predictions <- sapply(1:57,function(each.fam){  
  adjusted.p <- p.adjust(p = p.vals[[each.fam]],method = "fdr")
 #keep <- which(p.vals[[each.fam]] < .05)
  keep <- which(adjusted.p[[each.fam]] < .1)
  cor.mat <- cor(t(log2(asreml.counts[,keep] + 1)))

  test.fam <- rownames(fam.phenos)[each.fam]
  train.fam <- rownames(fam.phenos)[-each.fam]
  okriging(idtest = test.fam,idtrain = train.fam,corlist = list(cor.mat),H2vec = c(.99),pheno = fam.phenos,phenoname = "Volume")[,2] - mean(fam.phenos$Volume[-each.fam])
})

cor(the.predictions,true.val)^2
plot(the.predictions,true.val)
```

#### GLMNET

```{r}
suppressPackageStartupMessages(library(glmnet))

the.predictions <- sapply(1:57,function(each.fam){  
  these <- which(as.character(fam.phenos$fam_id) %in% unique(as.character(fam.phenos$fam_id))[each.fam])
  test.fam <- these
  train.fam <- c(1:190)[-these]
 #keep <- which(p.vals[[each.fam]] < .05)
  mean.cts <- apply(asreml.counts[-these,],2,mean)
sd.cts <- apply(asreml.counts[-these,],2,sd)
bcv.cts <- sd.cts/mean.cts
  keep <- which(bcv.cts > 1)
 gl.mod <- glmnet(x = log2(asreml.counts[-these,keep]),
                  y = fam.phenos$Volume[-these],alpha = 1) 
 out <- gl.mod$beta
 select <- which(out[,ncol(out)] > 0)

 cor.mat <- cor(t(log2(asreml.counts[,keep[select]] + 1)))
rownames(cor.mat) <- 1:190; colnames(cor.mat) <- 1:190
  these <- which(as.character(fam.phenos$fam_id) %in% unique(as.character(fam.phenos$fam_id))[each.fam])
  test.fam <- these
  train.fam <- c(1:190)[-these]
    mean(okriging(idtest = test.fam,idtrain = train.fam,corlist = list(cor.mat),H2vec = c(.99),pheno = fam.phenos,phenoname = "Volume")[,2] - mean(fam.phenos$Volume[-these]))
})


cor(the.predictions,true.val)^2
plot(the.predictions,true.val)
```
