## 1/27/18
*Raw mRNA sequencing libraries returned from the GSL are located at:

```linux
  /media/disk6/ARF/RNASEQ/rawreads/EWtarfiles && LGEPtarfiles
 ```
 
*Within the LGEP directory there are 18 .tar files (144 biological samples) and then an additional directory which includes tar files from the first run of sequencing.

*Within the EW directory there are 12 .tar files (96 biological samples)
  *The 12 tar files were delivered at two seperate times
     *First set included 4 tar files labeled: 20151203-GSL-046-B-Whetten-Pool9 && 10,11,12
     *Second set included 8 tar files labeled: GSL-047-A-Whetten-Pool1 && 2,3,4,5,6,7,8
  *The names of directories were manually edited to be lane01:lane12 corresponding to Pool  
      
###### 1. Untar sequencing lanes

```linux
    cd /media/disk6/ARF/RNASEQ/rawreads/EWtarfiles
    ls *.tar > EW.tar.list
    
    while read file; do tar -xf ./$file --directory ../EWfasta/; done < EW.tar.list
```    

```linux
    cd /media/disk6/ARF/RNASEQ/rawreads/LGEPtarfiles
    ls *.tar > LGEP.tar.list
    
    while read file; do tar -xf ./$file --directory ../LGEPfasta/; done < LGEP.tar.list
```    

###### 2. Run fastqc on raw files
      #Change into each of the 12 directories; create list of .fastq.gz files; make fastqc output dir and run fastqc
  
  ```linux 
   cd /media/disk6/ARF/RNASEQ/rawreads/EWfasta/lane10
   ls *.fastq.gz > lane10.list
   mkdir fastqc
   while read file; do /media/disk6/ARF/RNASEQ/software/FastQC/fastqc $file -o ./fastqc; done < lane10.list
   
   
   cd /media/disk6/ARF/RNASEQ/rawreads/EWfasta/lane12
   ls *.fastq.gz > lane12.list
   mkdir fastqc
   while read file; do /media/disk6/ARF/RNASEQ/software/FastQC/fastqc $file -o ./fastqc; done < lane12.list
    ...
    ...
   ```
    
    
***Noticed there is rRNA contamination in Sample 71 in EW lane 09.***


###### 3. Remove rRNA contamination

*Accessed www.arb-silva.de and searched for Pinus taeda

*805 Accessions showed up in results and were downloaded.  The original .tgz is located at 

```linux
/media/disk6/ARF/RNASEQ/software/arb-silva.de_2018-01-28_id495190.tgz
```

*The .fasta file inside the .tgz was placed in:

```linux
/media/disk6/ARF/RNASEQ/software/bbmap/resources
```

*bbDuk can be used to do overall quality trim, as well as trim first 10 bases, remove adapters, and matches to rRNA.

*Test it out on EW sample 84 from lane10
```linux
  ./software/bbmap/bbduk.sh -Xmx2g in=./rawreads/EWfasta/lane10/84_S29_L002_R1_001.fastq.gz out=./8429l2.fq.gz ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=1 ktrim=r k=25 mink=12 hdist=1 qtrim=rl trimq=20 minlength=50 maxlength=125
```
*Initially Sample 84 had rRNA and adapter contamination.  After running bbduk there was still one overresprestned sequence which when blasted hit to a tumor response protein.  Since this is biologicaly relevenat it should be OK to stay.  We can use bbduk now to process all raw reads.

*Article suggest low to minimium quality trimming works best: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4766705/
Also minimum read length for DE expression roughly 50 works: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4531809/

First run on the EW technical reps a single sample at a time
```linux
cd /media/disk6/ARF/RNASEQ
while read file; do ./software/bbmap/bbduk.sh -Xmx2g in=./rawreads/EWfasta/lane01/$file out=./trimmedfiltreads/EW/lane01/$file ref=./software/bbmap/resources/truseq.fa.gz,./software/bbmap/resources/rRNA_lob_pine.fasta forcetrimleft=10 t=20 ktrim=r k=25 mink=12 hdist=1 hdist2=0 qtrim=rl trimq=20 minlength=50 ; done < ./rawreads/EWfasta/lane01/lane01.list
```

*Output is located at:
```linux
/media/disk6/ARF/RNASEQ/trimmedfiltreads/EW/lane01 
```

All lanes are processed the same way by changing the lane numbers in the bbduk command.

